### PR TITLE
Filter stacktrace to just project frames

### PR DESCRIPTION
**Pull request by dpsutton**
Clojure stacktraces can often be quite difficult to read. This follows a recent addition to cider-nrepl which marked stackframes as coming from the project. This allows us to filter to the project stackframes.

The old filtering system is a "negative" filtering system. This would hide lines with the tags, and therefore our project would be hidden as it was marked with clj and repl and other tags. The solution is to add a "positive filter", ie, if a tag is found, show it, regardless if it overlaps with a negative filter as well.

So now in a stacktrace buffer, hit p to narrow down to just the project stakframes and then hit p again to go back to all stack frames.

normal-errors

project-errors-only

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines][1]
- [x] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][2] extremely useful.*

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md
[2]: https://cider.readthedocs.io/en/latest/hacking_on_cider/

### CODE REVIEW

File: cider-stacktrace.el [hunk 1]
```diff
@@ -550,6 +600,11 @@ prompt and whether to use a new window.  Similar to `cider-find-var'."
                           'help-echo (format "Toggle %s stack frames"
                                              (car filter)))
       (insert " "))
+    (insert-text-button "Project-Only"
```
Comment (xiongtx) [inline cider-stacktrace.el]: `Project` instead of `Project-Only` would be more in line with the naming of other options.

Comment (dpsutton) [inline cider-stacktrace.el]: The problem is that this filter works differently than the others. "Clojure" will remove or not remove all clojure namespace stackframes. This removes all _other_ frames but your project, which is why i named it Project-Only

Comment (xiongtx) [inline cider-stacktrace.el]: Oh, that makes sense.

Nonetheless I think `Project-Only` should be the first filter.

Comment (dpsutton) [inline cider-stacktrace.el]: moved it first. looks good there

**Comment by bbatsov** [comment]
The feature looks pretty great overall, but I'm worried people would be confused by it's different behaviour. Should we also have some version of this that actually fits into the current model of work? And how can we make it more obvious that this is intentionally behaving pretty differently from the rest of the filters?

**Comment by bbatsov** [comment]
Btw, I haven't read the code yet, but I assume that one you turn this on it turn off all the other filters, right?

**Comment by dpsutton** [comment]
> The feature looks pretty great overall, but I'm worried people would be confused by it's different behaviour. Should we also have some version of this that actually fits into the current model of work?

If you have one that mimics the current behavior of the other filters, it would remove all stackframes from the project, which seems almost useless to me, but it should be basically trivial to add.

> And how can we make it more obvious that this is intentionally behaving pretty differently from the rest of the filters?

I thought about that a little and it's kinda difficult. I haven't come up with anything good so I'm open to feedback or to user reports if there's confusion. Perhaps one thing we can do is to make the `project` tag exclusive to all others. Right now it overlaps so removing `repl` or `clj` can remove your project. But if project would "dominate" in that sense, perhaps it would behave like the others and we wouldn't need the "positive" filter. That notion only exists because we have to ensure the frames are shown when a filter says to remove it.

> Btw, I haven't read the code yet, but I assume that one you turn this on it turn off all the other filters, right?

Yeah, there's a defvar to hold the previous filters and i stick the current in there and then set all to filter and set the "positive" filters to `project.

One caveat is the `All` tag has a little bug in it in that it's not being underlined right now. I think there's some shenanigans because it's kind of a "pseudo" tag and not a real tag at all. The feature still works but only the underlining isn't flushed out. It's limited to the logic in determining what face to show in a little function so it should be easy to correct and test.

**Comment by bbatsov** [comment]
> If you have one that mimics the current behavior of the other filters, it would remove all stackframes from the project, which seems almost useless to me, but it should be basically trivial to add.

I guess you're right. We'll have to think about this down the road. 

> One caveat is the All tag has a little bug in it in that it's not being underlined right now. I think there's some shenanigans because it's kind of a "pseudo" tag and not a real tag at all. The feature still works but only the underlining isn't flushed out. It's limited to the logic in determining what face to show in a little function so it should be easy to correct and test.

Well, I guess we should improve this then. 

Overall the big goal for this feature would be to avoid confusion. Seems that we need some combination of turning off the other filters and a information message. Perhaps we can also add some configuration about the default filters for the stacktrace buffer.

**Comment by xiongtx** [comment]
The objection, as I understand it, is that `Project-Only` will hide *non-project* stack frames, whereas other filters only show/hide frames they name.

I don't think it's too confusing to have a `Project-Only` filter, but perhaps it'd be sufficient to have a `Project` filter that behaves like the other filters.

**Comment by bbatsov** [comment]
> I don't think it's too confusing to have a Project-Only filter, but perhaps it'd be sufficient to have a Project filter that behaves like the other filters.

Yeah, that's what I thought initially myself.

**Comment by dpsutton** [comment]
We could make the project tag exclusive of all others. The problem with this approach is that you probably only want to see project which means you want every _other_ filter than project. So for the project filter to be meaningful it would need to really set the others.

If this is the preferred way I'm happy to do that but I think it makes the project button rather irrelevant. As its setup now, on the stacktrace just hit `p` and it shows project only, and then hit `p` again and it restores the previous view.

If we prefer the standard filter the tags must be made mutually exclusive as the `project` tag overlaps with the other tags.

**Comment by dpsutton** [comment]
> Some dependencies were not available: spinner, queue
> make: *** [elpa] Error 255
> The command "make elpa" failed and exited with 2 during .

Transient bug in the CI. They pass for me. I'll make a minor change and push later to retrigger but otherwise i think we are good here.

**Comment by bbatsov** [comment]
Seems you forgot to have a separate commit for the actual project filter. It would also be nice if the commit messages complied with the project's commit message style.

**Comment by dpsutton** [comment]
I can update that tonight

**Comment by bbatsov** [comment]
OK

**Comment by vspinu** [comment]
How about a separator before ALL to emphasize that ALL  and Project-Only have different semantics?

```
  Show: Clojure Java REPL Tooling Duplicates  |  All Project-Only  (21 frames hidden)
```

**Comment by bbatsov** [comment]
> How about a separator before ALL to emphasize that ALL and Project-Only have different semantics?

I like the idea.

In the commit messages:

- "Only show" -> "Show only"
- "Adds" -> "Add"
- "Introduces" -> "Introduce"

**Comment by dpsutton** [comment]
with-separator

A separator like this?

@bbatsov sorry about the present tense. I thought that this project liked that but then i read the document about commit messages by tpope after. I'll update that now

**Comment by dpsutton** [comment]
or with earmuffs
with-earmuffs

**Comment by bbatsov** [comment]
I think the idea was to have both "All" and "Project Only" in a separate section (separated with a vertical bar for instance), as pressing them eradicates all the other filters so in a way they are similar/special.

**Comment by dpsutton** [comment]
Gotcha.  It sounds silly but that will require some more programming. The "(11 frames hidden)" serves to separate the top of this buffer from the bottom. I started getting rid of some of this questionable stuff with `cider-stacktrace-frame-p` rather than it hiding and unhiding every line of text on the buffer.

**Comment by bbatsov** [comment]
Got it. Well - let's make it better then!

**Comment by dpsutton** [comment]
showing-normal

showing-project-only
I think this UI is working for me. The other filters kinda don't show anything they allow you to hide (negative filter). Whereas the others are show project-only or show all. Seems to indicate that they are distinct mechanisms.

**Comment by dpsutton** [comment]
although I've inverted the underlining to make it clear which are hidden now

indicate-hiding

**Comment by bbatsov** [comment]
Looks good to me! 👍 Thanks for all the work you've put into this!

**Comment by vspinu** [comment]
The show/hide semantics is unfortunate IMO. Is it the underlined keys the one which is shown or hidden? Show/hide suggests that those two lines are opposite of each other, which is obviously not the case.

"Hide" makes things disappear, it's a negative concept, so if you press on a key (an "yes" action) you have to think about which stacks will disappear. It's confusing. Pressing on "yes" with "show" and thinking about which things appear instead is much more straightforward.

How about this:

```
  Show: Project-Only All 
  Type: Java REPL Tooling Duplicates  (25 frames hidden)
```   

"Type" means type of the stacktrace entry to show. It will also align as "type" has same number of letters as "show".

BTW, when clicking on `Project-Only` it is not getting underlined for me.

**Comment by vspinu** [comment]
Could be also 

```
  Filter:  Project-Only 
  Show: Clojure Java REPL Tooling Duplicates All  (19 frames hidden)
```
Entirely backward compatible.

**Comment by dpsutton** [comment]
ah, yes, you're right about the underlining with Project only. I was always using the `p` shortcut which works correctly. I'll figure that one out.

As to my thinking about the show/hide naming, I started from a stacktrace having a bunch of frames in it I don't want to see, so I suppress some of them. The underline means that those types are hidden. But like all things, I've been looking at it for a while so if it looks strange to you I'm inclined to accept that. I have no strong opinions one way or the other so if there's confusion I'm happy to change it.

**Pull request merged**